### ISSUE TITLE
Perl 5.8.0 segfault

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6188]
Migrated from [rt.perl.org#19566](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=19566) (status was 'resolved')

Searchable as RT19566$

**Comment by p5pRT** [comment]
### From pfedor@fuw.edu.pl


Hello\,

I am using Red Hat Linux 8\.0 and Perl 5\.8\.0\. I have stumbled onto 
something that I think might be a bug in Perl&#8203;:

pamela&#8203;:\~/smiecie/segfault2$ cat segfault\.pl
\#\!/usr/bin/perl 

my $l = \[ \{'a' => 1\} \];
push @&#8203;$l\, \{'a' => 1\};
for \(@&#8203;$l\) \{
&nbsp;        while \(\<DUPA>\) \{\} 
\}
pamela&#8203;:\~/smiecie/segfault2$ \./segfault\.pl 
Segmentation fault
pamela&#8203;:\~/smiecie/segfault2$

Here is the stack trace from gdb 
\(/home/caesar/builds/perl/install/bin/perl is just Perl 5\.8\.0 compiled 
for debugging\)&#8203;:

pamela&#8203;:\~/smiecie/segfault2$ gdb /home/caesar/builds/perl/install/bin/perl
GNU gdb Red Hat Linux \(5\.2\.1\-4\)
Copyright 2002 Free Software Foundation\, Inc\.
GDB is free software\, covered by the GNU General Public License\, and you are
welcome to change it and/or distribute copies of it under certain conditions\.
Type "show copying" to see the conditions\.
There is absolutely no warranty for GDB\.  Type "show warranty" for details\.
This GDB was configured as "i386\-redhat\-linux"\.\.\.
\(gdb\) r segfault\.pl 
Starting program&#8203;: /home/caesar/builds/perl/install/bin/perl segfault\.pl

Program received signal SIGSEGV\, Segmentation fault\.
0x42074a1a in free \(\) from /lib/i686/libc\.so\.6
\(gdb\) bt
\#0  0x42074a1a in free \(\) from /lib/i686/libc\.so\.6
\#1  0x080aa966 in Perl\_safesysfree \(where=0x815cd74\) at util\.c&#8203;:151
\#2  0x080d02bc in Perl\_sv\_clear \(sv=0x816675c\) at sv\.c&#8203;:5049
\#3  0x080d069c in Perl\_sv\_free \(sv=0x816675c\) at sv\.c&#8203;:5192
\#4  0x080e8daf in Perl\_free\_tmps \(\) at scope\.c&#8203;:181
\#5  0x0805ff3b in perl\_run \(my\_perl=0x815cae8\) at perl\.c&#8203;:1606
\#6  0x0805c9ac in main \(argc=2\, argv=0xbffff9a4\, env=0xbffff9b0\)
&nbsp;    at perlmain\.c&#8203;:85
\#7  0x420158d4 in \_\_libc\_start\_main \(\) from /lib/i686/libc\.so\.6
\(gdb\)

And here is what valgrind said&#8203;:

pamela&#8203;:\~/smiecie/segfault2$ valgrind /home/caesar/builds/perl/install/bin/perl segfault\.pl
==9294== valgrind\-1\.0\.4\, a memory error detector for x86 GNU/Linux\.
==9294== Copyright \(C\) 2000\-2002\, and GNU GPL'd\, by Julian Seward\.
==9294== Estimated CPU clock rate is 350 MHz
==9294== For more details\, rerun with&#8203;: \-v
==9294== 
==9294== Invalid free\(\) / delete / delete\[\]
==9294==    at 0x4003BCE8&#8203;: free \(vg\_clientfuncs\.c&#8203;:180\)
==9294==    by 0x80AA966&#8203;: Perl\_safesysfree \(util\.c&#8203;:151\)
==9294==    by 0x80D02BC&#8203;: Perl\_sv\_clear \(sv\.c&#8203;:5049\)
==9294==    by 0x80D069C&#8203;: Perl\_sv\_free \(sv\.c&#8203;:5192\)
==9294==    Address 0x40A482D4 is 636 bytes inside a block of size 1008 alloc'd
==9294==    at 0x4003BA4E&#8203;: malloc \(vg\_clientfuncs\.c&#8203;:100\)
==9294==    by 0x80AA637&#8203;: Perl\_safesysmalloc \(util\.c&#8203;:78\)
==9294==    by 0x80C7333&#8203;: S\_more\_sv \(sv\.c&#8203;:271\)
==9294==    by 0x80CEE26&#8203;: Perl\_newSV \(sv\.c&#8203;:4422\)
==9294== 
==9294== ERROR SUMMARY&#8203;: 1 errors from 1 contexts \(suppressed&#8203;: 0 from 0\)
==9294== malloc/free&#8203;: in use at exit&#8203;: 48482 bytes in 556 blocks\.
==9294== malloc/free&#8203;: 705 allocs\, 150 frees\, 61112 bytes allocated\.
==9294== For a detailed leak analysis\,  rerun with&#8203;: \-\-leak\-check=yes
==9294== For counts of detected errors\, rerun with&#8203;: \-v
pamela&#8203;:\~/smiecie/segfault2$ 

With best regards

Aleksander

"There's more to see than can ever be seen
 More to do than can ever be done"

**Comment by p5pRT** [comment]
### From @nwc10

On Sun\, Dec 29\, 2002 at 05&#8203;:41&#8203;:41PM \-0000\, Pawel Aleksander Fedorynski wrote&#8203;:
> I am using Red Hat Linux 8\.0 and Perl 5\.8\.0\. I have stumbled onto 
> something that I think might be a bug in Perl&#8203;:

\(Roughly\) if perl goes segfault\, it's a bug\.

> pamela&#8203;:\~/smiecie/segfault2$ cat segfault\.pl
> \#\!/usr/bin/perl 
> 
> my $l = \[ \{'a' => 1\} \];
> push @&#8203;$l\, \{'a' => 1\};
> for \(@&#8203;$l\) \{
>         while \(\<DUPA>\) \{\} 
> \}
> pamela&#8203;:\~/smiecie/segfault2$ \./segfault\.pl 
> Segmentation fault

It seems to be the re\-use of $\_ while it contains any sort of reference

The smallest examples I can get it down to are&#8203;:

\#\!/usr/bin/perl

for \(\\$a\) \{
&nbsp;  1 while \<DUPA>;
\}

\#\!/usr/bin/perl

for \(\[\]\) \{
&nbsp;  1 while \<DUPA>;
\}

\#\!/usr/bin/perl

for \(\{\}\) \{
&nbsp;  1 while \<DUPA>;
\}

all of which segfault for me on Debian\. On FreeBSD I see&#8203;:

\./perl \-Ilib ~/test/segfault\.pl 
perl in free\(\)&#8203;: warning&#8203;: modified \(chunk\-\) pointer

This is still happening in the development perl \(patch 18378\)

> And here is what valgrind said&#8203;:
> 
> pamela&#8203;:\~/smiecie/segfault2$ valgrind /home/caesar/builds/perl/install/bin/perl segfault\.pl
> ==9294== valgrind\-1\.0\.4\, a memory error detector for x86 GNU/Linux\.
> ==9294== Copyright \(C\) 2000\-2002\, and GNU GPL'd\, by Julian Seward\.
> ==9294== Estimated CPU clock rate is 350 MHz
> ==9294== For more details\, rerun with&#8203;: \-v
> ==9294== 
> ==9294== Invalid free\(\) / delete / delete\[\]
> ==9294==    at 0x4003BCE8&#8203;: free \(vg\_clientfuncs\.c&#8203;:180\)
> ==9294==    by 0x80AA966&#8203;: Perl\_safesysfree \(util\.c&#8203;:151\)
> ==9294==    by 0x80D02BC&#8203;: Perl\_sv\_clear \(sv\.c&#8203;:5049\)
> ==9294==    by 0x80D069C&#8203;: Perl\_sv\_free \(sv\.c&#8203;:5192\)
> ==9294==    Address 0x40A482D4 is 636 bytes inside a block of size 1008 alloc'd
> ==9294==    at 0x4003BA4E&#8203;: malloc \(vg\_clientfuncs\.c&#8203;:100\)
> ==9294==    by 0x80AA637&#8203;: Perl\_safesysmalloc \(util\.c&#8203;:78\)
> ==9294==    by 0x80C7333&#8203;: S\_more\_sv \(sv\.c&#8203;:271\)
> ==9294==    by 0x80CEE26&#8203;: Perl\_newSV \(sv\.c&#8203;:4422\)
> ==9294== 
> ==9294== ERROR SUMMARY&#8203;: 1 errors from 1 contexts \(suppressed&#8203;: 0 from 0\)
> ==9294== malloc/free&#8203;: in use at exit&#8203;: 48482 bytes in 556 blocks\.
> ==9294== malloc/free&#8203;: 705 allocs\, 150 frees\, 61112 bytes allocated\.
> ==9294== For a detailed leak analysis\,  rerun with&#8203;: \-\-leak\-check=yes
> ==9294== For counts of detected errors\, rerun with&#8203;: \-v
> pamela&#8203;:\~/smiecie/segfault2$ 

Thanks\. That ought to be really helpful\, \(because it's what I'd do next if
you hadn't already done it\) but even with that it's not obvious to me what
is going wrong\.

\(For p5p&#8203;: S\_more\_sv seems to be allocating more arena space\. So it's not clear
why Perl\_sv\_free is being called at all for the pointer later on\. Solve that
and we probably have the cause\)

Nicholas Clark

**Comment by p5pRT** [comment]
### From smcc@mit.edu

>>>>> "NC" == Nicholas Clark \<nick@&#8203;unfortu\.net> writes&#8203;:

NC> On Sun\, Dec 29\, 2002 at 05&#8203;:41&#8203;:41PM \-0000\, Pawel Aleksander Fedorynski wrote&#8203;:
>> I am using Red Hat Linux 8\.0 and Perl 5\.8\.0\. I have stumbled onto 
>> something that I think might be a bug in Perl&#8203;:

NC> \(Roughly\) if perl goes segfault\, it's a bug\.

PAF> pamela&#8203;:\~/smiecie/segfault2$ cat segfault\.pl

PAF> \#\!/usr/bin/perl 
PAF> 
PAF> my $l = \[ \{'a' => 1\} \];
PAF> push @&#8203;$l\, \{'a' => 1\};
PAF> for \(@&#8203;$l\) \{
PAF>     while \(\<DUPA>\) \{\} 
PAF> \}

PAF> pamela&#8203;:\~/smiecie/segfault2$ \./segfault\.pl 
PAF> Segmentation fault

NC> It seems to be the re\-use of $\_ while it contains any sort of
NC> reference

The only kind of reuse that causes the problem\, though\, is getting the
result of a read\-line \(\<FH>\) from a closed filehandle\.

NC> The smallest examples I can get it down to are&#8203;:

NC> \#\!/usr/bin/perl

NC> for \(\\$a\) \{
NC>   1 while \<DUPA>;
NC> \}

NC> \.\.\.

NC> all of which segfault for me on Debian\. On FreeBSD I see&#8203;:

NC> \./perl \-Ilib ~/test/segfault\.pl 
NC> perl in free\(\)&#8203;: warning&#8203;: modified \(chunk\-\) pointer

NC> This is still happening in the development perl \(patch 18378\)

My favorite short example&#8203;:

% \./perl \-we 'for \(\\2\) \{ $\_ = \<FH> \}'
Name "main&#8203;::FH" used only once&#8203;: possible typo at \-e line 1\.
readline\(\) on unopened filehandle FH at \-e line 1\.
zsh&#8203;: segmentation fault  \./perl \-we 'for \(\\2\) \{ $\_ = \<FH> \}'

PAF> And here is what valgrind said&#8203;:
PAF> 
PAF> ==9294== Invalid free\(\) / delete / delete\[\]
PAF> ==9294==    at 0x4003BCE8&#8203;: free \(vg\_clientfuncs\.c&#8203;:180\)
PAF> ==9294==    by 0x80AA966&#8203;: Perl\_safesysfree \(util\.c&#8203;:151\)
PAF> ==9294==    by 0x80D02BC&#8203;: Perl\_sv\_clear \(sv\.c&#8203;:5049\)
PAF> ==9294==    by 0x80D069C&#8203;: Perl\_sv\_free \(sv\.c&#8203;:5192\)
PAF> ==9294==    Address 0x40A482D4 is 636 bytes inside a block of size 1008 alloc'd
PAF> ==9294==    at 0x4003BA4E&#8203;: malloc \(vg\_clientfuncs\.c&#8203;:100\)
PAF> ==9294==    by 0x80AA637&#8203;: Perl\_safesysmalloc \(util\.c&#8203;:78\)
PAF> ==9294==    by 0x80C7333&#8203;: S\_more\_sv \(sv\.c&#8203;:271\)
PAF> ==9294==    by 0x80CEE26&#8203;: Perl\_newSV \(sv\.c&#8203;:4422\)

NC> Thanks\. That ought to be really helpful\, \(because it's what I'd do
NC> next if you hadn't already done it\) but even with that it's not
NC> obvious to me what is going wrong\.

NC> \(For p5p&#8203;: S\_more\_sv seems to be allocating more arena space\. So
NC> it's not clear why Perl\_sv\_free is being called at all for the
NC> pointer later on\. Solve that and we probably have the cause\)

Exactly\. valgrind shows that were trying to free a pointer to an SV
using free\(\)\, which is wrong because SVs aren't allocated individually
with malloc\(\); they're allocated in 1008\-byte arenas that are only
supposed to be freed all at once\. The fact that references are
involved suggests a way this sort of confusion could happen&#8203;: when an
SV holds a reference\, the pointer that would normally point to
malloc\(\)ed string points instead to the pointed\-to SV\.

The other key observation is that the bug doesn't happen when a string
is read from the filehandle\, only in the special case when you get
"undef" from a closed handle&#8203;:

% echo "okay" | \./perl \-e 'for \(\\$a\) \{ $\_ = \<> \}'
\[no segfault\]

Also\, if the "\\$a" is a "\\2" instead\, most ways you would try to
modify $\_ croak about trying to modify a read\-only value\. The readline
operator is funny in that for efficiency\, it modifies the SV you're
reading into in place\, rather than with copying\. \(The '$x = \<FH>'
syntax is actually special\-cased to eliminate the assignment
operator\)\. A problem occurs in the special case of yielding undef\,
since it does the equivalent of \`undef' on its target SV without first
checking that the SV isn't readonly\, releasing its reference\, etc\.

The following patch also fixes a similar problem that looks like it
was waiting to happen in the branch later in the function that handles
certain EOF conditions\, though I couldn't reproduce a problem there\.

 \-\- Stephen

<details><summary>Inline Patch</summary>

```diff
--- pp_hot.c.orig	2003-01-02 17:00:01.000000000 -0500
+++ pp_hot.c	2003-01-02 17:51:51.000000000 -0500
@@ -1466,6 +1466,8 @@ Perl_do_readline(pTHX)
 		report_evil_fh(PL_last_in_gv, io, PL_op->op_type);
 	}
 	if (gimme == G_SCALAR) {
+	    /* undef TARG, and push that undefined value */
+	    SV_CHECK_THINKFIRST_COW_DROP(TARG);
 	    (void)SvOK_off(TARG);
 	    PUSHTARG;
 	}
@@ -1527,6 +1529,7 @@ Perl_do_readline(pTHX)
 		}
 	    }
 	    if (gimme == G_SCALAR) {
+		SV_CHECK_THINKFIRST_COW_DROP(TARG);
 		(void)SvOK_off(TARG);
 		SPAGAIN;
 		PUSHTARG;
```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Thu\, Jan 02\, 2003 at 02&#8203;:33&#8203;:14PM \+0000\, Nicholas Clark wrote&#8203;:
> It seems to be the re\-use of $\_ while it contains any sort of reference
> 
> The smallest examples I can get it down to are&#8203;:
> 
> \#\!/usr/bin/perl
> 
> for \(\\$a\) \{
>   1 while \<DUPA>;
> \}

I think it is a combination of two problematic things\.

First\, in pp\_hot\.c&#8203;:Perl\_do\_readline\(\)\, line 1469\, where the target is set
to undef on failure\, the code simply does a

&nbsp;    \(void\)SvOK\_off\(TARG\);

This is wrong\, as when $\_ is an RV\, the referent is silently leaked\.
It also means we have an RV with \!SvROK\, which leads to\.\.\.

Second\, in sv\_clear\(\)\, the code for freeing a RV looks a bit like&#8203;:

&nbsp;    case SVt\_PV&#8203;:
&nbsp;    case SVt\_RV&#8203;:
&nbsp;    if \(SvROK\(sv\)\) \{
&nbsp;        \.\.\.
&nbsp;    \}
&nbsp;    else if \(SvPVX\(sv\) && SvLEN\(sv\)\)
&nbsp;        \.\.\.

Thus\, for an RV whose SvROK has been turned off\, non\-existent
SvPVX etc fields are accessed and cleared\, hence the coredump\.

I'm not sure how to fix the first problem \- there doesn't seem
to be an API function that does the equivalent of 'undef';
pp\_undef\(\) does it itself rather than calling a general\-purpose function\.
So\, folks\, what's the right way to undef an SV?

In the second case\, that basic code structure has been around
since 5\.000\, so\, I'm really puzzled how this hasn't caused a problem
before\. Why are the SVt\_PV and SVt\_RV cases treated the same?
Is there something more subtle that I'm not seeing here?

Dave\.

\-\- 
A walk of a thousand miles begins with a single step\.\.\.
then continues for another 1\,999\,999 or so\.

**Comment by p5pRT** [comment]
### From @nwc10

On Thu\, Jan 02\, 2003 at 11&#8203;:07&#8203;:46PM \+0000\, Dave Mitchell wrote&#8203;:
> In the second case\, that basic code structure has been around
> since 5\.000\, so\, I'm really puzzled how this hasn't caused a problem
> before\. Why are the SVt\_PV and SVt\_RV cases treated the same?
> Is there something more subtle that I'm not seeing here?

Here's the part I can answer quickly\, using Stephen McCamant's excellent
1 liner&#8203;:

$ perl5\.00503 \-we 'for \(\\2\) \{ $\_ = \<FH> \}'
Name "main&#8203;::FH" used only once&#8203;: possible typo at \-e line 1\.
Read on closed filehandle \<FH> at \-e line 1\.
Segmentation fault
$ perl5\.00404 \-we 'for \(\\2\) \{ $\_ = \<FH> \}'
Name "main&#8203;::FH" used only once&#8203;: possible typo at \-e line 1\.
Read on closed filehandle \<FH> at \-e line 1\.
Segmentation fault

It's a good old bug\, that strangely no\-one has found before\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From smcc@mit.edu

>>>>> "DM" == Dave Mitchell \<davem@&#8203;fdgroup\.com> writes&#8203;:

DM> On Thu\, Jan 02\, 2003 at 02&#8203;:33&#8203;:14PM \+0000\, Nicholas Clark wrote&#8203;:
NC> It seems to be the re\-use of $\_ while it contains any sort of
NC> reference

NC> The smallest examples I can get it down to are&#8203;:

NC> \#\!/usr/bin/perl
NC> 
NC> for \(\\$a\) \{
NC> 1 while \<DUPA>;
NC> \}

DM> I think it is a combination of two problematic things\.

DM> First\, in pp\_hot\.c&#8203;:Perl\_do\_readline\(\)\, line 1469\, where the target
DM> is set to undef on failure\, the code simply does a

DM>     \(void\)SvOK\_off\(TARG\);

DM> This is wrong\, as when $\_ is an RV\, the referent is silently
DM> leaked\.  It also means we have an RV with \!SvROK\, which leads
DM> to\.\.\.

The way I would describe the problematic condition is that we have an
SV with SvRV/SvPVX set\, but ROK is off and the pointer doesn't point
to a malloc\(\)ed buffer\.

DM> Second\, in sv\_clear\(\)\, the code for freeing a RV looks a bit like&#8203;:

DM>     case SVt\_PV&#8203;:
DM>     case SVt\_RV&#8203;:
DM>     if \(SvROK\(sv\)\) \{
DM>         \.\.\.
DM>     \}
DM>     else if \(SvPVX\(sv\) && SvLEN\(sv\)\)
DM>         \.\.\.

DM> Thus\, for an RV whose SvROK has been turned off\, non\-existent
DM> SvPVX etc fields are accessed and cleared\, hence the coredump\.

Actually\, the SvPVX field is still pointing to another SV&#8203;: accessing
the pointer isn't a problem\, but passing it to free\(\) causes free\(\) to
segfault \(for some free\(\) implementaions\)\, since the SV wasn't
allocated \(individually\) by malloc\(\)\.

DM> I'm not sure how to fix the first problem \- there doesn't seem to
DM> be an API function that does the equivalent of 'undef'; pp\_undef\(\)
DM> does it itself rather than calling a general\-purpose function\.
DM> So\, folks\, what's the right way to undef an SV?

It think it depends on what exact effect you want to have\, and what
kind of SV you're working with\. Most of the complexity of pp\_undef has
to do with SVs holding non\-scalar data like arrays and code refs\.
Undef on globs is also a special case\. Two other questions are whether
you want to free the SV's buffer\, and whether you want to invoke its
set magic if it has any\. As you can see from my patch\, the only parts
of undef\-ing that I think are really important are checking the
THINKFIRST flags \(which you always need to do before changing an SV\)\,
and the SvOK\_off\(\)\, which invalidates any numeric or string contents\.
If you also want to do glob assignment correctly\, you need to go
through sv\_setsv\(\) \(perhaps through a macro\); if the destination isn't
a glob\, "sv\_setsv\(sv\, &sv\_undef\)" basically does a THINKFIRST check
and then an SvOK\_off\. If you also want to call set magic\, you need to
do that too\.

\(This implies that in the version with the patch I sent earlier\,
"\*glob = \<CLOSED\_FH>" and "$tied = \<CLOSED\_FH>" will still act
subtly differently than "\*glob = undef" and "$tied = undef"\, by not
giving the warning about "undefined value assigned to typeglob" and by
not calling the STORE method\, respectively\. Similarly for other magic
like "$\#x = \<CLOSED\_FH>"\, "$\< = \<CLOSED\_FH>"\, etc\.\)

DM> In the second case\, that basic code structure has been around
DM> since 5\.000\, so\, I'm really puzzled how this hasn't caused a
DM> problem before\. Why are the SVt\_PV and SVt\_RV cases treated the
DM> same?  Is there something more subtle that I'm not seeing here?

\* The PVX field and the RV field are the same field\. An RV has just
&nbsp;  the PVX/RV field\, while a PV also has CUR and LEN\. When an SV is
&nbsp;  upgraded to PV or above and then stores a reference again\, the
&nbsp;  reference is stored in the "PVX" field\, but ROK is set\. This is why
&nbsp;  PVs\, PVIVs\, PVNVs\, PVMGs\, and PVBMs all have to go through the
&nbsp;  "if \(SvROK\(sv\)\)" check\.

\* When a reference becomes not a reference \(in sv\_unref\(\)\)\, along with
&nbsp;  ROK being cleared\, the PVX/RV field is zeroed out \(before
&nbsp;  potentially being replaced with the stringified representation of
&nbsp;  the old reference\, if the reference was converted into a string\)\.

Roughly speaking\, then\, there are three states an SV can be in&#8203;:

\* ROK is set\, and PVX/RV points to another SV\.

\* ROK is off\, but PVX/RV is non null\, in which case it's a pointer to
&nbsp;  a malloc\(\)ed buffer\.

\* PVX/RV is null\.

Note that if ROK is off\, the code in sv\_clear is careful not to touch
CUR or LEN unless PVX is non\-null\, since the SV might be a undef'ed RV
in which case the CUR and LEN fields would be off the end of the
structure\. As long as the RV field is cleared whenever ROK is\, this is
safe\, which is why except for this bug that code hasn't been a
problem\.

 \-\- Stephen

**Comment by p5pRT** [comment]
### From @iabyn

On Thu\, Jan 02\, 2003 at 09&#8203;:46&#8203;:13PM \-0500\, Stephen McCamant wrote&#8203;:

\[lots of useful stuff\]

Thanks for the explanation \- I posted my 'I think this is this problem'
email before seeing your patch\. I'm now happy to say that your patch
addresses any concerns I may have had :\-\)

\-\- 
My get\-up\-and\-go just got up and went\.

**Comment by p5pRT** [comment]
### From @abigail

On Thu\, Jan 02\, 2003 at 11&#8203;:23&#8203;:12PM \+0000\, Nicholas Clark wrote&#8203;:
> On Thu\, Jan 02\, 2003 at 11&#8203;:07&#8203;:46PM \+0000\, Dave Mitchell wrote&#8203;:
> > In the second case\, that basic code structure has been around
> > since 5\.000\, so\, I'm really puzzled how this hasn't caused a problem
> > before\. Why are the SVt\_PV and SVt\_RV cases treated the same?
> > Is there something more subtle that I'm not seeing here?
> 
> Here's the part I can answer quickly\, using Stephen McCamant's excellent
> 1 liner&#8203;:
> 
> $ perl5\.00503 \-we 'for \(\\2\) \{ $\_ = \<FH> \}'
> Name "main&#8203;::FH" used only once&#8203;: possible typo at \-e line 1\.
> Read on closed filehandle \<FH> at \-e line 1\.
> Segmentation fault
> $ perl5\.00404 \-we 'for \(\\2\) \{ $\_ = \<FH> \}'
> Name "main&#8203;::FH" used only once&#8203;: possible typo at \-e line 1\.
> Read on closed filehandle \<FH> at \-e line 1\.
> Segmentation fault
> 
> It's a good old bug\, that strangely no\-one has found before\.


Introduced in 5\.003\_01&#8203;:

$ /opt/perl/5\.003/bin/perl \-we 'for \(\\2\) \{ $\_ = \<FH> \}'  
Identifier "main&#8203;::FH" used only once&#8203;: possible typo at \-e line 1\.
Read on closed filehandle \<FH> at \-e line 1\.
$ /opt/perl/5\.003\_01/bin/perl \-we 'for \(\\2\) \{ $\_ = \<FH> \}'  
Name "main&#8203;::FH" used only once&#8203;: possible typo at \-e line 1\.
Read on closed filehandle \<FH> at \-e line 1\.
Segmentation fault
$


Abigail

**Comment by p5pRT** [comment]
### From arthur@contiller.se


On fredag\, jan 3\, 2003\, at 00&#8203;:07 Europe/Stockholm\, Dave Mitchell wrote&#8203;:

> I'm not sure how to fix the first problem \- there doesn't seem
> to be an API function that does the equivalent of 'undef';
> pp\_undef\(\) does it itself rather than calling a general\-purpose 
> function\.
> So\, folks\, what's the right way to undef an SV?

This might already be answered\, but I think the correct way is to

sv\_setsv\(\*sv\,&PL\_sv\_undef\);

Arthur

**Comment by p5pRT** [comment]
### From @rgs

Stephen McCamant wrote&#8203;:
> The following patch also fixes a similar problem that looks like it
> was waiting to happen in the branch later in the function that handles
> certain EOF conditions\, though I couldn't reproduce a problem there\.

Thanks\, applied to bleadperl as \#18485\, including a test case
by Jarkko\.

**Comment by p5pRT** [comment]
### From @jhi

Since the problem was tracked down and patched\, I'm marking the problem
ticket as resolved\.

**Comment by p5pRT** [comment]
@jhi - Status changed from 'new' to 'resolved'
